app/src/main/java/com/aaronjwood/portauthority/utils/MDNSResolver.java

```diff
@@ -0,0 +1,80 @@
+package com.aaronjwood.portauthority.utils;
+
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.InetAddress;
+import java.net.MulticastSocket;
+
+public class MDNSResolver implements Closeable {
+	InetAddress mdnsIP = InetAddress.getByName("224.0.0.251");
+	private int mdnsPort = 5353;
+	MulticastSocket socket = new MulticastSocket();
+
+	public MDNSResolver(int timeout) throws IOException {
+		socket.setSoTimeout(timeout);
+		socket.setTimeToLive(1);
+		// TODO: iPhones respond only back to the multicast address and port 5353 provided requestId == 0
+		socket.joinGroup(mdnsIP);
+	}
+
+	void writeName(DataOutputStream out, String name) throws IOException {
+		int s = 0, e;
+		while ((e = name.indexOf('.', s)) != -1) {
+			out.writeByte(e - s);
+			out.write(name.substring(s, e).getBytes());
+			s = e + 1;
+		}
+		out.write(name.length() - s);
+		out.write(name.substring(s).getBytes());
+		out.writeByte(0);
+	}
+
+	String decodeName(byte[] data, int offset, int length) {
+		StringBuilder s = new StringBuilder(length);
+		for (int i = offset; i < offset + length; i++) {
+			byte len = data[i];
+			if (len == 0) break;
+			s.append(new String(data, i + 1, len)).append('.');
+			i += len;
+		}
+		s.setLength(s.length() - 1);
+		return s.toString();
+	}
+
+	byte[] dnsRequest(int id, String name) throws IOException {
+		ByteArrayOutputStream baos = new ByteArrayOutputStream();
+		DataOutputStream out = new DataOutputStream(baos);
+		out.writeShort(id);
+		out.write(new byte[] {0, 0, 0, 1, 0, 0, 0, 0, 0, 0});
+		writeName(out, name);
+		out.write(new byte[] {0, 0xc, 0, 1});
+		return baos.toByteArray();
+	}
+
+	String reverseName(byte[] addr) {
+		// note: only IPv4 is supported here
```

Comment:
Currently the host scanning picks up on IPv6 stuff. At least on my phone in my home network it does due to the way I have my LAN configured. It looks like we will have to disable IPv6 completely due to this limitation.